dashboard/test/models/pd/application/teacher1819_application_test.rb

```diff
@@ -349,5 +350,82 @@ class Teacher1819ApplicationTest < ActiveSupport::TestCase
+    end
+
+    test 'find_default_workshop finds a teachercon workshop for applications with a G3 partner' do
+      teachercon_workshops = {}
+      [Pd::Workshop::COURSE_CSD, Pd::Workshop::COURSE_CSP].each do |course|
+        TEACHERCONS.each do |teachercon|
+          city = teachercon[:city]
+          teachercon_workshops[[course, city]] = create :pd_workshop,
+            num_sessions: 1, course: course, subject: Pd::Workshop::SUBJECT_TEACHER_CON, location_address: city
+        end
+      end
+
+      g3_partner_name = REGIONAL_PARTNER_TC_MAPPING.keys.sample
+      g3_partner = build :regional_partner, group: 3, name: g3_partner_name
+      application = build :pd_teacher1819_application, regional_partner: g3_partner
+
+      [Pd::Workshop::COURSE_CSD, Pd::Workshop::COURSE_CSP].each do |course|
+        city = get_matching_teachercon(g3_partner)[:city]
+        workshop = teachercon_workshops[[course, city]]
+
+        application.course = course === Pd::Workshop::COURSE_CSD ? 'csd' : 'csp'
+        assert_equal workshop, application.find_default_workshop
+      end
+    end
+
+    test 'find_default_workshop find an appropriate partner workshop for G1 and G2 partners' do
+      program_manager = create :workshop_organizer
+      partner = create :regional_partner
+      create :regional_partner_program_manager,
+        program_manager: program_manager,
+        regional_partner: partner
+
+      # where "appropriate workshop" is the earliest teachercon, fit, or local
+      # summer workshop matching the application course.
+
+      invalid_workshop = create :pd_workshop, organizer: program_manager
+      create :pd_session,
+        workshop: invalid_workshop,
+        start: Date.new(2018, 1, 10)
+
+      earliest_valid_workshop = create :pd_workshop, :local_summer_workshop, organizer: program_manager
+      create :pd_session,
+        workshop: earliest_valid_workshop,
+        start: Date.new(2018, 1, 15)
+
+      latest_valid_workshop = create :pd_workshop, :local_summer_workshop, organizer: program_manager
+      create :pd_session,
+        workshop: latest_valid_workshop,
+        start: Date.new(2018, 12, 15)
+
+      application = build :pd_teacher1819_application, course: 'csp', regional_partner: partner
+      assert_equal earliest_valid_workshop, application.find_default_workshop
+    end
+
+    test 'setting and changing pd_workshop_id automatically enrolls user' do
+      school_info = create :school_info
+      user = create :teacher, school_info: school_info
+      application = create :pd_teacher1819_application, user: user
+      first_workshop = create :pd_workshop
+      second_workshop = create :pd_workshop
+
+      application.pd_workshop_id = first_workshop.id
+
+      assert_creates(Pd::Enrollment) do
+        application.save!
+      end
+
+      application.pd_workshop_id = second_workshop.id
+
+      assert_creates(Pd::Enrollment) do
+        application.save!
+      end
```

Comment:
Does this also unenroll from the previous workshop? Can we assert that too? If not, it probably should.

Reply:
Should it? On the one hand, that does prevent the situation in which we accidentally enroll the user in multiple courses. But on the other hand, what if an application gets accidentally updated to a workshop in which the user is already enrolled?

Reply:
Ideally it would be idempotent in that case. How about something along these lines:

 1. Unenroll if the previous workshop is different
 2. Enroll with `find_or_create_by!` so it won't create duplicates

Reply:
Sorry, I wasn't clear: if the application gets accidentally updated to a workshop in which the user is already enrolled, it will as currently implemented not create a duplicate. BUT since the update was an accident, we would presumably want to change it away to a different workshop, at which point the user would get unenrolled from their legitimate workshop.

Reply:
Why would we want the assigned workshop to not match the enrolled workshop? In your example, wouldn't we want the "legitimate" workshop to be assigned and enrolled? If we manually change assigned to a different workshop, then wouldn't that become the new legitimate workshop (assigned and enrolled)?

Reply:
Scenario:

- Teacher is enrolled in some workshop A that is entirely unrelated to teachercon stuff
- Teacher's application is accidentally updated to reference workshop A. Teacher is already enrolled in workshop, so no enrollment is created
- Mistake is noticed, and teacher's application is then updated to reference the correct teachercon workshop. Teacher's enrollment in workshop A is deleted

I don't like the idea of being able to delete teacher's enrollments out from under them.

Reply:
That's fair. However I also don't like the idea of littering our system with enrollments whenever the assigned workshop is changed. That also seems undesirable.

We can only remove the previously assigned workshop enrollment, rather than any. Note the [changes](https://apidock.com/rails/ActiveModel/Dirty/changes) hash will contain the old and new values (at least for the properties hash as a whole if not the individual property).

I agree we shouldn't remove any unrelated enrollments out from under the teacher, but the partners are supposed to manage the 5-day local summer workshop enrollment, and we're supposed to manage teachercon enrollment, so in those cases of assigned workshops I'd argue it does make sense.

Alternatively, we could forgo automatic enrollment on the fly and instead have a manual step of enrolling everyone at once in their assigned workshop once it's settled.

Thoughts?

Reply:
Oops, forgot this was still open.

I'm down for a manual enrollment step; is there a place in the ui that would fit without too much extra work?